src/sync.h

```diff
@@ -400,4 +400,98 @@ class CSemaphoreGrant
     }
 };
 
+/**
+ * A smart pointer that protects the owned object with a mutex.
+ * Every time the pointer is dereferenced a mutex is acquired and held locked
+ * for the duration of the method being executed. For example, the following is
+ * safe:
+ * @code{.cpp}
+ * ThreadSafePtr<std::vector<int>> v;
+ *
+ * std::thread t1{[&v]() { v->push_back(5); }};
+ * std::thread t2{[&v]() { v->push_back(6); }};
+ * @endcode
+ *
+ * It is possible so acquire the mutex for longer time than just for a single call:
+ * @code{.cpp}
+ * ThreadSafePtr<std::vector<int>> v{5, 6, 7, 8};
+ *
+ * {
+ *     auto v_locked = *v;  // acquires the mutex
+ *     for (auto& i : v_locked) {
+ *         i *= 10;
+ *     }
+ *     // the mutex is released when `v_locked` is destroyed
+ * }
+ * // will result in `v` being `{50, 60, 70, 80}`
+ * @endcode
+ */
+template <typename T>
+class ThreadSafePtr
+{
+public:
+    /**
+     * Construct the smart pointer and its internal object of type T,
+     * passing `args...` to its constructor.
+     */
+    template <typename... Args>
+    ThreadSafePtr(Args... args) : m_obj{args...}
+    {
+    }
+
+    /**
+     * An auxiliary class that holds a mutex locked for the duration of its
+     * lifetime and provides access to another object. For exaple:
+     * @code{.cpp}
+     * Proxy p{mutex, obj}; // acquires the mutex
+     *
+     * p->foo(); // `foo()` is a method of `obj`, the mutex is still locked
+     * p->bar(); // `bar()` is a method of `obj`, the mutex is still locked
+     *
+     * // if `obj` provides `operator[]` (e.g. `std::map`), then the following is also ok:
+     * p[5] = 10; // the mutex is still locked
+     *
+     * // if `obj` provides `begin()`/`end()` methods, then the following is also ok:
+     * for (auto& x : p) { ... } // the mutex is still locked
+     *
+     * // the mutex is released when `p` is destroyed
+     * @endcode
+     */
+    class Proxy
+    {
+    public:
+        Proxy(Mutex& mutex, T& obj)
+            : m_lock{mutex, "ThreadSafePtrMutex", __FILE__, __LINE__},
+              m_raw_ptr_to_obj{&obj}
+        {
+        }
+
+        T* operator->() { return m_raw_ptr_to_obj; }
+        const T* operator->() const { return m_raw_ptr_to_obj; }
+
+        template <typename Key>
+        auto& operator[](Key key)
+        {
+            return (*m_raw_ptr_to_obj)[key];
+        }
+
+        auto begin() { return m_raw_ptr_to_obj->begin(); }
+        auto end() { return m_raw_ptr_to_obj->end(); }
+
+    private:
+        const UniqueLock<Mutex> m_lock;
+        T* const m_raw_ptr_to_obj;
+    };
+
+    Proxy operator->() { return Proxy{m_mutex, m_obj}; }
+    const Proxy operator->() const { return Proxy{m_mutex, m_obj}; }
```

Comment:
I removed them because they are not needed (currently). In a previous incarnation of this I had it working with `const ThreadSafePtr`, but removed the complications that were required for that.

Can always extend it (restore these) if `const` objects are necessary.